google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryImpl.java

```diff
@@ -1184,9 +1225,81 @@ public Boolean call() {
   public TableResult query(QueryJobConfiguration configuration, JobOption... options)
       throws InterruptedException, JobException {
     Job.checkNotDryRun(configuration, "query");
+
+    // If all parameters passed in configuration are supported by the query() method on the backend,
+    // put on fast path
+    QueryRequestInfo requestInfo = new QueryRequestInfo(configuration);
+    if (requestInfo.isFastQuerySupported()) {
+      String projectId = getOptions().getProjectId();
+      QueryRequest content = requestInfo.toPb();
+      return queryRpc(projectId, content, options);
+    }
+    // Otherwise, fall back to the existing create query job logic
     return create(JobInfo.of(configuration), options).getQueryResults();
   }
 
+  private TableResult queryRpc(
+      final String projectId, final QueryRequest content, JobOption... options) {
+    com.google.api.services.bigquery.model.QueryResponse results;
+    try {
+      results =
+          runWithRetries(
+              new Callable<com.google.api.services.bigquery.model.QueryResponse>() {
+                @Override
+                public com.google.api.services.bigquery.model.QueryResponse call() {
+                  return bigQueryRpc.queryRpc(projectId, content);
+                }
+              },
+              getOptions().getRetrySettings(),
+              EXCEPTION_HANDLER,
+              getOptions().getClock());
+    } catch (RetryHelperException e) {
+      throw BigQueryException.translateAndThrow(e);
+    }
+
+    if (results.getErrors() != null) {
+      List<BigQueryError> bigQueryErrors =
+          Lists.transform(results.getErrors(), BigQueryError.FROM_PB_FUNCTION);
+      // Throwing BigQueryException since there may be no JobId and we want to stay consistent
+      // with the case where there there is a HTTP error
+      throw new BigQueryException(bigQueryErrors);
+    }
+
+    Schema schema = results.getSchema() == null ? null : Schema.fromPb(results.getSchema());
+    Long numRows;
+    if (results.getNumDmlAffectedRows() == null && results.getTotalRows() == null) {
+      numRows = 0L;
+    } else if (results.getNumDmlAffectedRows() != null) {
+      numRows = results.getNumDmlAffectedRows();
+    } else {
+      numRows = results.getTotalRows().longValue();
+    }
+
+    if (results.getPageToken() != null) {
+      JobId jobId = JobId.fromPb(results.getJobReference());
+      boolean jobStatus = results.getJobComplete();
+      String cursor = results.getPageToken();
+      return new TableResult(
+          schema,
+          numRows,
+          new PageImpl<>(
+              // fetch next pages of results
+              new QueryPageFetcher(
+                  jobId, jobStatus, schema, getOptions(), cursor, optionMap(options)),
+              cursor,
+              // cache first page of result
+              transformTableData(results.getRows(), schema)));
+    }
```

Comment:
Is results.getRows()

There is special casing in the getQueryResults path here:

https://github.com/googleapis/java-bigquery/blob/b3f59b1e1215d93a3b7048149c88318f2a074f79/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/Job.java#L304

Looking at the change, it looks like it should work. Is there a test? I know I might have mentioned about DDL test before. Just double checking.

Reply:
Yes we are handling DDL differently here as well: https://github.com/googleapis/java-bigquery/pull/509/files#diff-71270aede5da05bfa317bb9a38f082e0R1270 

Confirming DDL tests were added before: [integration test](https://github.com/googleapis/java-bigquery/pull/509/files#diff-af63ea32c5525b76f3837bc5aa8e5a25R1581), [retry mock test](https://github.com/googleapis/java-bigquery/pull/509/files#diff-ecb4841d47a85a6afc924d6d942b95b0R2226)